go/cmd/dolt/commands/checkout.go

```diff
@@ -56,59 +56,81 @@ func Checkout(ctx context.Context, commandStr string, args []string, dEnv *env.D
 		usagePrt()
 		return 1
+	}
+
+	var verr errhand.VerboseError
+	if apr.ContainsArg(doltdb.DocTableName) {
+		verr = errhand.BuildDError("Use dolt checkout to checkout individual docs.").Build()
+		cli.PrintErrln(verr.Verbose())
+		return 1
+	}
+
+	if newBranch, newBranchOk := apr.GetValue(coBranchArg); newBranchOk {
+		return checkoutNewBranch(ctx, dEnv, newBranch, apr)
+	}
+
+	name := apr.Arg(0)
+
+	if isBranch, err := actions.IsBranch(ctx, dEnv, name); err != nil {
+		verr = errhand.BuildDError("error: unable to determine type of checkout").AddCause(err).Build()
+		cli.PrintErrln(verr.Verbose())
+		return 1
+	} else if isBranch {
+		return checkoutKnownBranch(ctx, dEnv, name)
+	}
+
+	tbls, docs, err := actions.GetTblsAndDocDetails(dEnv, args)
+	if err != nil {
+		verr = errhand.BuildDError("error: unable to parse arguments.").AddCause(err).Build()
+		cli.PrintErrln(verr.Verbose())
+		return 1
+	}
+
+	verr = checkoutTablesAndDocs(ctx, dEnv, tbls, docs)
+
+	if verr != nil && apr.NArg() == 1 {
+		verr = tryCheckoutNewBranch(ctx, dEnv, name)
+	}
+
+	if verr != nil {
+		cli.PrintErrln(verr.Verbose())
+		return 1
+	}
+
+	return 0
+}
+
+func tryCheckoutNewBranch(ctx context.Context, dEnv *env.DoltEnv, name string) errhand.VerboseError {
+	var verr errhand.VerboseError
+
+	if ref, refExists, err := getRef(ctx, dEnv, name); err != nil {
+		verr = errhand.BuildDError("fatal: unable to read from data repository.").AddCause(err).Build()
+	} else if refExists {
+		verr = checkoutNewBranchFromStartPt(ctx, dEnv, name, ref)
 	} else {
-		var verr errhand.VerboseError
-		newBranch, nbOk := apr.GetValue(coBranchArg)
+		verr = errhand.BuildDError("error: could not find %s", name).Build()
+	}
 
-		if nbOk {
-			startPt := "head"
-			if apr.NArg() == 1 {
-				startPt = apr.Arg(0)
-			}
+	return verr
+}
 
-			verr = checkoutNewBranch(ctx, dEnv, newBranch, startPt)
-		} else {
-			name := apr.Arg(0)
-			isBranch, rootsWithTable, err := actions.BranchOrTable(ctx, dEnv, name)
-
-			if err != nil {
-				verr = errhand.BuildDError("fatal: unable to read from data repository.").AddCause(err).Build()
-			} else if !rootsWithTable.IsEmpty() || apr.NArg() > 1 {
-				verr = checkoutTable(ctx, dEnv, apr.Args())
-			} else if isBranch {
-				verr = checkoutBranch(ctx, dEnv, name)
-			} else {
-				refs, err := dEnv.DoltDB.GetRefs(ctx)
-
-				if err != nil {
-					verr = errhand.BuildDError("fatal: unable to read from data repository.").AddCause(err).Build()
-				}
-
-				found := false
-				for _, rf := range refs {
-					if remRef, ok := rf.(ref.RemoteRef); ok && remRef.GetBranch() == name {
-						verr = checkoutNewBranch(ctx, dEnv, name, rf.String())
-						found = true
-						break
-					}
-				}
-
-				if !found {
-					verr = errhand.BuildDError("error: could not find %s", name).Build()
-				}
-			}
-		}
+func getRef(ctx context.Context, dEnv *env.DoltEnv, name string) (string, bool, error) {
```

Comment:
Why does this return a string instead of a ref?

Reply:
Also would call this getRemoteBranchRef

Reply:
Renamed, and returning `ref` now instead of `ref.String()`: https://github.com/liquidata-inc/dolt/pull/299/files#diff-78081aa9ea2b0558c0f56aad2a236961R109